### ISSUE TITLE
Influence model cannot execute functional call

### ISSUE THREAD

**Issue by danielkaplan137** [aai-institute/pyDVL#640]
While using CgInfluence on a pre-trained CNN I get the following error from torch:
`    raise RuntimeError("The stateless API can't be used with Jitted modules")`

with the full backtrace:
`influences = influence_model.influences(x, y, x,y, mode="up")
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/anaconda3/envs/sci/lib/python3.12/site-packages/pydvl/influence/base_influence_function_model.py", line 162, in influences
    return self._influences(x_test, y_test, x, y, mode)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/opt/anaconda3/envs/sci/lib/python3.12/site-packages/pydvl/influence/base_influence_function_model.py", line 453, in _influences
    return cast(TensorType, sum(tensors))
                            ^^^^^^^^^^^^

  File "/opt/anaconda3/envs/sci/lib/python3.12/site-packages/pydvl/influence/types.py", line 631, in generate_interactions
    yield comp_block.interactions(left_batch, right_batch, mode)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/opt/anaconda3/envs/sci/lib/python3.12/site-packages/pydvl/influence/types.py", line 431, in interactions
    return bilinear_form.grads_inner_prod(left_batch, right_batch, self.gp)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/opt/anaconda3/envs/sci/lib/python3.12/site-packages/pydvl/influence/types.py", line 253, in grads_inner_prod
    left_grad = gradient_provider.flat_grads(left)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/opt/anaconda3/envs/sci/lib/python3.12/site-packages/pydvl/influence/torch/base.py", line 251, in flat_grads
    self.grads(batch).values(), shape=(batch.x.shape[0], -1)
    ^^^^^^^^^^^^^^^^^

  File "/opt/anaconda3/envs/sci/lib/python3.12/site-packages/pydvl/influence/torch/base.py", line 190, in grads
    gradient_dict = self._grads(batch.to(self.device))
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/opt/anaconda3/envs/sci/lib/python3.12/site-packages/pydvl/influence/torch/base.py", line 120, in _grads
    result: Dict[str, torch.Tensor] = torch.vmap(
                                      ^^^^^^^^^^^

  File "/opt/anaconda3/envs/sci/lib/python3.12/site-packages/torch/_functorch/apis.py", line 203, in wrapped
    return vmap_impl(
           ^^^^^^^^^^

  File "/opt/anaconda3/envs/sci/lib/python3.12/site-packages/torch/_functorch/vmap.py", line 331, in vmap_impl
    return _flat_vmap(
           ^^^^^^^^^^^

  File "/opt/anaconda3/envs/sci/lib/python3.12/site-packages/torch/_functorch/vmap.py", line 479, in _flat_vmap
    batched_outputs = func(*batched_inputs, **kwargs)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/opt/anaconda3/envs/sci/lib/python3.12/site-packages/torch/_functorch/apis.py", line 399, in wrapper
    return eager_transforms.grad_impl(func, argnums, has_aux, args, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/opt/anaconda3/envs/sci/lib/python3.12/site-packages/torch/_functorch/eager_transforms.py", line 1449, in grad_impl
    results = grad_and_value_impl(func, argnums, has_aux, args, kwargs)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/opt/anaconda3/envs/sci/lib/python3.12/site-packages/torch/_functorch/vmap.py", line 48, in fn
    return f(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^

  File "/opt/anaconda3/envs/sci/lib/python3.12/site-packages/torch/_functorch/eager_transforms.py", line 1407, in grad_and_value_impl
    output = func(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^

  File "/opt/anaconda3/envs/sci/lib/python3.12/site-packages/pydvl/influence/torch/base.py", line 116, in _compute_loss
    outputs = functional_call(self.model, params, (x.unsqueeze(0).to(self.device),))
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/opt/anaconda3/envs/sci/lib/python3.12/site-packages/torch/_functorch/functional_call.py", line 148, in functional_call
    return nn.utils.stateless._functional_call(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^`

**Comment by mdbenito** [comment]
Can you provide a bit more context to reproduce the issue? A minimal working example with your CNN. Also, which version of pydvl are you using?

@schroedk Any ideas?

**Comment by schroedk** [comment]
Hi @danielkaplan137,
Thank you for reporting this issue. Do you have access to the source code of the model and the ability to run a version that is not JIT-compiled?
It seems that the restriction arises from the torch.func stateless API, which is not compatible with JIT-compiled models. Nevertheless, providing a minimal working example would be helpful to determine if modifying the implementation is straightforward.
Thanks:)

**Comment by danielkaplan137** [comment]
@mdbenito @schroedk 
Thanks so much for responding so swiftly guys. I'm using the latest devel version, i.e., `'0.9.3.dev0'`

My network structure is very simple as well. It contains a backbone and two branches.

class Net(nn.Module):
        def __init__(self):
            super(Net, self).__init__()

#BackBone
            self.seq1 = nn.Sequential(
                nn.Flatten(),
                nn.Linear(120, 64),
                nn.ReLU()
                )

#branch 1
            self.seq2 = nn.Sequential(
                nn.Linear(64, 128),
                nn.ReLU(),
                nn.Linear(128, 512),
                nn.ReLU(),
                nn.Dropout(0.25),
                nn.Linear(512, 128),
                nn.ReLU(),
                nn.Dropout(0.5),
                nn.Linear(128, 2),
                nn.Softmax(dim=1)
            )

#Branch 2
            self.seq3 = nn.Sequential(
                nn.Linear(64, 128),
                nn.ReLU(),
                nn.Linear(128, 512),
                nn.ReLU(),
                nn.Dropout(0.25),
                nn.Linear(512, 128),
                nn.ReLU(),
                nn.Dropout(0.5),
                nn.ReLU(),
                nn.Linear(128, 1)
            )

        def forward(self, x):
            x = self.seq1(x)
            output1 = self.seq2(x)
            output2 = self.seq3(x)
            return [output2, output1]

I'm not sure which part is jitted? All of them? Anything I can do within my already trained networks? 
Happy to hear what can be done! Thanks!

**Comment by danielkaplan137** [comment]
Here's a simple code to reproduce

[export.zip](https://github.com/user-attachments/files/18480729/export.zip)

**Comment by schroedk** [comment]
Hi @danielkaplan137, I am sorry for the long delay. Unfortunately, it is not straight-forward to use the torch.func stateless api for jitted models. For now, you could use a non-jitted version of your model, in combination with a faster influence value approximation, which could look like this (using the latest development version)
```python
import torch
import logging
import torch.nn as nn
from torch.utils.data import DataLoader

from pydvl.influence.torch import InverseHarmonicMeanInfluence, BlockMode

dev=torch.device("cpu")
pretrained_net= torch.jit.load('./model_scripted.pt')
state_dict = pretrained_net.state_dict()
train_ds=torch.load('./train_ds.pt')
pretrained_net.double()
pretrained_net.to(dev)
loss =torch.nn.MSELoss(reduction='sum')

class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()

        #BackBone
        self.seq1 = nn.Sequential(
            nn.Flatten(),
            nn.Linear(120, 64),
            nn.ReLU()
        )

        #branch 1
        self.seq2 = nn.Sequential(
            nn.Linear(64, 128),
            nn.ReLU(),
            nn.Linear(128, 512),
            nn.ReLU(),
            nn.Dropout(0.25),
            nn.Linear(512, 128),
            nn.ReLU(),
            nn.Dropout(0.5),
            nn.Linear(128, 2),
            nn.Softmax(dim=1)
        )

        #Branch 2
        self.seq3 = nn.Sequential(
            nn.Linear(64, 128),
            nn.ReLU(),
            nn.Linear(128, 512),
            nn.ReLU(),
            nn.Dropout(0.25),
            nn.Linear(512, 128),
            nn.ReLU(),
            nn.Dropout(0.5),
            nn.ReLU(),
            nn.Linear(128, 1)
        )

    def forward(self, x):
        x = self.seq1(x)
        output1 = self.seq2(x)
        output2 = self.seq3(x)
        return [output2, output1]


def loss_fn(pred,target):
    yb_g_split_1 = (target[:, 2]).to(dev) #Tc
    #pred = pretrained_net(input)
    return loss(pred[0].flatten(), yb_g_split_1.flatten())


batch_size = 64
train_loader = DataLoader(train_ds, batch_size, drop_last=True, shuffle=False, num_workers=0)

non_scripted_model = Net().double().to(dev)
non_scripted_model.load_state_dict(state_dict)


if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO)
    x,y = next(iter(train_loader))
    test_loader = DataLoader(train_ds, batch_size, drop_last=True, shuffle=False, num_workers=0)
    model = non_scripted_model
    model.eval()

    influence_model = InverseHarmonicMeanInfluence(model, loss_fn, 0.001, block_structure=BlockMode.LAYER_WISE)
    influence_model = influence_model.fit(train_loader)
    values = influence_model.influences(x, y)


```
